idaes/apps/grid_integration/examples/thermal_generator_storage.py

```diff
@@ -0,0 +1,1069 @@
+    _add_UT_DT_constraints(m)
+
+    ## Expression
+    # energy generated by all the units
+    def total_power_fun(m,j,h):
+        return m.P_total[j,h] == m.P_R[j,h] + m.P_D[j,h] - m. P_C[j,h]
+    m.tot_power = pyo.Constraint(m.UNITS,m.HOUR,rule = total_power_fun)
+
+    def prod_cost_fun(m,j,h):
+        return m.min_load_cost[j] * m.on_off[j,h] \
+        + sum(m.F[j,l]*m.power_segment[j,h,l] for l in m.SEGMENTS)
+    m.prod_cost_approx = pyo.Expression(m.UNITS,m.HOUR,rule = prod_cost_fun)
+
+    # start up costs
+    def start_cost_fun(m,j,h):
+        return m.start_up_cost[j]*m.start_up[j,h]
+    m.start_up_cost_expr = pyo.Expression(m.UNITS,m.HOUR,rule = start_cost_fun)
+
+    # total cost
+    def tot_cost_fun(m,j,h):
+        return m.prod_cost_approx[j,h] + m.start_up_cost_expr[j,h]
+    m.tot_cost = pyo.Expression(m.UNITS,m.HOUR,rule = tot_cost_fun)
+
+    ## Objective
+    def exp_revenue_fun(m):
+        return sum(m.P_total[j,h]*m.DAM_price[h]- m.tot_cost[j,h] for h in m.HOUR for j in m.UNITS)
+    m.exp_revenue = pyo.Objective(rule = exp_revenue_fun,sense = pyo.maximize)
+    m.exp_revenue.deactivate()
+
+    return m
+
+def _add_tracking_constraints_and_objectives(m, ramping_weight = 20):
+
+    # add params for objective weights
+    m.cost_weight = pyo.Param(initialize = 1, mutable = False)
+    m.ramping_weight = pyo.Param(initialize = ramping_weight, mutable = False)
+
+    ## USE L1 norm
+    # add a slack variable for ramping
+    m.slack_var_ramp = pyo.Var(m.UNITS,m.HOUR,within = pyo.NonNegativeReals)
+
+    # add constraints for ramping
+    def abs_ramp_con_fun1(m,j,h):
+        if h == 0:
+            return -m.slack_var_ramp[j,h]<=m.P_T[j,h] - m.pre_P_T[j]
+        else:
+            return -m.slack_var_ramp[j,h]<=m.P_T[j,h] - m.P_T[j,h-1]
+    m.abs_ramp_con1 = pyo.Constraint(m.UNITS,m.HOUR,rule = abs_ramp_con_fun1)
+
+    def abs_ramp_con_fun2(m,j,h):
+        if h == 0:
+            return m.slack_var_ramp[j,h]>=m.P_T[j,h] - m.pre_P_T[j]
+        else:
+            return m.slack_var_ramp[j,h]>=m.P_T[j,h] - m.P_T[j,h-1]
+    m.abs_ramp_con2 = pyo.Constraint(m.UNITS,m.HOUR,rule = abs_ramp_con_fun2)
+
+    def power_abs_least(m):
+        return m.ramping_weight * sum(m.slack_var_ramp[j,h] for h in m.HOUR for j in m.UNITS)\
+               + m.cost_weight * sum(m.tot_cost[j,h] for h in m.HOUR for j in m.UNITS)
+    m.abs_dev = pyo.Objective(rule = power_abs_least,sense = pyo.minimize)
+
+    return
+
+def build_tracking_model(model_data,plan_horizon,ramping_weight = 20,segment_number=4,model_mode = '1a'):
+
+    m = _build_model(model_data,plan_horizon,segment_number)
+
+    _add_tracking_constraints_and_objectives(m, ramping_weight = ramping_weight)
+    _add_slack_vars(m)
+    _add_slack_constraints(m)
+    _add_min_slack_objective(m)
+    _add_slack_bounding_constraints(m)
+
+    switch_model_mode(m,mode = model_mode)
+
+    return m
+
+def switch_model_mode(m,mode = '1a'):
+
```

Comment:
Missing doc string.

Reply:
I recommend making the modes more descriptive. `no-storage-market-interactions` instead of `1a`.